### ISSUE TITLE
Some images are only displayed in black & white, old Intel (before 2009) and QEMU-emulated CPUs

### ISSUE THREAD

**Issue by Zharkan** [lovell/sharp#3893]
## Possible install-time or require-time problem

- [x] I have read the [documentation relating to installation](https://sharp.pixelplumbing.com/install).

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

### Are you using a supported runtime?

- [x] I am using Node.js with a version that satisfies `^18.17.0 || ^20.3.0 || >=21.0.0`

### Are you using a supported package manager and installing optional dependencies?

- [x] I am using npm >= 9.6.5 with `--include=optional`

### What is the complete output of running `npm install --verbose --foreground-scripts sharp` in an empty directory?

<details>

```
npm info using npm@10.2.5
npm info using node@v20.10.0
npm verb title npm install sharp
npm verb argv "install" "--loglevel" "verbose" "--foreground-scripts" "sharp"
npm verb logfile logs-max:10 dir:/home/media_gestion/.npm/_logs/2023-12-11T23_18_28_144Z-
npm verb logfile /home/media_gestion/.npm/_logs/2023-12-11T23_18_28_144Z-debug-0.log
npm http fetch GET 200 https://registry.npmjs.org/sharp 353ms (cache revalidated)
npm verb reify failed optional dependency /path/to/project/node_modules/fsevents
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-win32-x64-msvc
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-win32-ia32-msvc
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-win32-arm64-msvc
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-linux-arm64-musl
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-linux-arm64-gnu
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-darwin-x64
npm verb reify failed optional dependency /path/to/project/node_modules/@next/swc-darwin-arm64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-win32-x64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-win32-ia32
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-wasm32
npm verb reify failed optional dependency /path/to/project/node_modules/@emnapi/runtime
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-linuxmusl-arm64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-linux-s390x
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-linux-arm64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-linux-arm
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-libvips-linuxmusl-arm64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-libvips-linux-s390x
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-libvips-linux-arm64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-libvips-linux-arm
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-libvips-darwin-x64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-libvips-darwin-arm64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-darwin-x64
npm verb reify failed optional dependency /path/to/project/node_modules/@img/sharp-darwin-arm64
npm http fetch POST 200 https://registry.npmjs.org/-/npm/v1/security/advisories/bulk 308ms
```

</details>

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```
  System:
    OS: Linux 5.15 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
    CPU: (8) x64 Common KVM processor
    Memory: 2.10 GB / 5.16 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 20.10.0 - /usr/bin/node
    Yarn: 1.22.19 - ~/.npm-global/bin/yarn
    npm: 10.2.5 - ~/.npm-global/bin/npm
  npmPackages:
    sharp: ^0.33.0 => 0.33.0 
```

I'm building my project with Sharp (Next.js app) and some images are black & white. There's also a message when I'm running `npm run build` : 

```
 ✓ Linting and checking validity of types    
   ▲ Next.js 14.0.4
   - Environments: .env.production
   - Experiments (use at your own risk):
     · scrollRestoration

   Creating an optimized production build  ...WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
   Creating an optimized production build  .WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
```

**Comment by lovell** [comment]
> WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000

> CPU: (8) x64 Common KVM processor

This is due to the use of a custom CPU, possibly QEMU-emulated, with somewhat non-standard flags.

There's an upstream fix to support this configuration at https://github.com/libvips/libvips/pull/3763 that has not yet been included in a libvips release.

You can workaround it for now by ensuring the hypervisor allows the use of CPU flags such as SSE4, or by setting the `VIPS_NOVECTOR` environment variable to switch off the use of SIMD.

**Comment by icetee** [comment]
> > WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
> 
> > CPU: (8) x64 Common KVM processor
> 
> This is due to the use of a custom CPU, possibly QEMU-emulated, with somewhat non-standard flags.
> 
> There's an upstream fix to support this configuration at [libvips/libvips#3763](https://github.com/libvips/libvips/pull/3763) that has not yet been included in a libvips release.
> 
> You can workaround it for now with one of the following:
> 
> * ensuring the hypervisor allows the use of CPU flags such as SSE4
> * setting the `VIPS_NOVECTOR` environment variable
> * using `sharp.simd(false)` to switch off the use of SIMD.
> 
> https://sharp.pixelplumbing.com/api-utility#simd

The `VIPS_NOVECTOR` env is not working, but `sharp.simd(false)` did.

**Comment by lovell** [comment]
@icetee You'll need to set `VIPS_NOVECTOR` before calling `require('sharp')` (I've updated my comment to reflect this).

**Comment by icetee** [comment]
I will use Docker and I defined the env before installing enithink.

If print the envs (printenv), I see VIPS_NOVECTOR=0

This is correct environment :/

**Comment by kleisauke** [comment]
It looks like the `VIPS_NOVECTOR` environment variable is not effective due to this line:
https://github.com/lovell/sharp/blob/c9e39960075ef84ba604896ada1c318d000a7bdf/lib/utility.js#L204

**Comment by lovell** [comment]
D'oh! I'll take a look at your PR now @icetee, thank you.

**Comment by AndrStnz** [comment]
> > WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
> 
> > CPU: (8) x64 Common KVM processor
> 
> This is due to the use of a custom CPU, possibly QEMU-emulated, with somewhat non-standard flags.
> 
> There's an upstream fix to support this configuration at [libvips/libvips#3763](https://github.com/libvips/libvips/pull/3763) that has not yet been included in a libvips release.
> 
> You can workaround it for now with one of the following:
> 
> * ensuring the hypervisor allows the use of CPU intrinsics such as SSE4

Setting the CPU Type in Proxmox to "Host" instead of "default" did indeed solve the issue for me.
Thank you very much for the hint!

**Comment by lovell** [comment]
v0.33.2 now available with prebuilt binaries that include libvips v8.15.1 and therefore the upstream fix.

**Comment by ImamAlfariziSyahputra** [comment]
sorry, how to fix this `WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000`? in nextjs with docker. from this conversation i don't reallly get it how to fix it, can someone help? thanks!

**Comment by lovell** [comment]
@ImamAlfariziSyahputra Please ensure you are using at least version 0.33.2 of sharp.

**Comment by ImamAlfariziSyahputra** [comment]
> @ImamAlfariziSyahputra Please ensure you are using at least version 0.33.2 of sharp.

I am using the latest version:
![image](https://github.com/lovell/sharp/assets/63343105/3edd3706-a8d1-4ae8-803b-bd58bb40c23c)

The warning console:
![image](https://github.com/lovell/sharp/assets/63343105/0f34edc2-e26a-4c04-8121-2434078f8050)

**Comment by kachkaev** [comment]
Same here inside a docker container on Linux:

```
#22 16.36 WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
(repeats many times)
```

```
pnpm why sharp

dependencies:
image-minimizer-webpack-plugin 3.8.3
└── sharp 0.33.2 peer
sharp 0.33.2
```

**Comment by lovell** [comment]
If you're relying on a globally-installed libvips, please ensure it is at least version v8.15.1.

**Comment by kleisauke** [comment]
The following warning:
```
WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000
```

[originates from Highway](https://github.com/google/highway/blob/1.2.0/hwy/targets.cc#L707-L716) and means there's a mismatch between what your CPU supports and what the prebuilt binaries expects. Specifically, the [`x86::DetectTargets()`](https://github.com/google/highway/blob/1.2.0/hwy/targets.cc#L689) function detects that your CPU supports only SSE2, [a baseline feature of x64 architectures](https://github.com/google/highway/blob/1.2.0/hwy/targets.cc#L336). However, the prebuilt binaries were compiled targeting the [`x86-64-v2` micro-architecture](https://github.com/lovell/sharp-libvips/pull/155), which corresponds to [Highway's S-SSE3 feature 'group'](https://github.com/google/highway/blob/1.2.0/hwy/targets.cc#L287-L288).

<details>
  <summary>Details</summary>

```c++
#include <cstdio>
#include <cstdint>

#define HWY_EMU128 (1LL << 61)
#define HWY_SCALAR (1LL << 62)

#define HWY_SSE2 (1LL << 14)
#define HWY_SSSE3 (1LL << 12)  // S-SSE3

int main() {
  const uint64_t bits_u = static_cast<uint64_t>(HWY_SCALAR | HWY_EMU128 | HWY_SSE2);
  const uint64_t enabled = static_cast<uint64_t>(HWY_SCALAR | HWY_SSSE3 | HWY_SSE2);
  fprintf(stderr,
          "WARNING: CPU supports 0x%08x%08x, software requires 0x%08x%08x\n",
          static_cast<uint32_t>(bits_u >> 32),
          static_cast<uint32_t>(bits_u & 0xFFFFFFFF),
          static_cast<uint32_t>(enabled >> 32),
          static_cast<uint32_t>(enabled & 0xFFFFFFFF));
  // WARNING: CPU supports 0x6000000000004000, software requires 0x4000000000005000

  return 0;
}
```
</details>

TL;DR: this warning on an `x86-64` system likely means the `ssse3` flag is missing when the CPU is queried with the `CPUID` instruction. This can happen for two reasons:
1. Your CPU is too old and doesn't support S-SSE3, which could cause a crash (`SIGILL`) if the software tries to use those instructions (see also #2723).
2. You're running inside a virtual machine, and the hypervisor is masking the `CPUID` feature flags of the underlying CPU, which would affect performance.

To check what features your CPU supports on Linux, you can run:
```console
$ cat /proc/cpuinfo
```

**Comment by Aluisio** [comment]
Hello @lovell , I still have the same problem. In my case, any version after version 0.32.6 shows this warning (I tested it on all stable versions).

Is there any solution or compatibility check?

**Comment by kleisauke** [comment]
@Aluisio I've just updated the comment above yours to remove the complex (LLM-generated) wording. Please open [a new installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) and answer all of the questions if you still require help.
